### PR TITLE
Feature/unit tests

### LINKED ISSUE #60: Add Unit Tests For Basic Request, Response, and Notification RPCs

### PR DESCRIPTION
**Pull request by russjohnson09**
Fixes #60 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
```npm run test``` script added which will run tests under the tests/node directory. Right now there are basic examples of request, response, and notification rpcs.

### Summary
Basic request, response, and notification rpc tests copied over from the java suite sdk and adjusted to work with mocha instead of JUnit.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) and this pull request adheres to the [Contributing Guide](./CONTRIBUTING.md).

### CODE REVIEW

File: tests/node/rpc/structs/ImageTests.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+const Test = require('./../../../Test.js');
+const Validator = require('./../../../Validator.js');
+const assertTrue = Validator.assertTrue.bind(Validator);
+const assertEquals = Validator.assertEquals.bind(Validator);
+const assertNull = Validator.assertNull.bind(Validator);
+const assertNullOrUndefined = Validator.assertNullOrUndefined.bind(Validator);
+const assertNotNull = Validator.assertNotNull.bind(Validator);
+
+let msg;
```
Comment (crokita) [inline tests/node/rpc/structs/ImageTests.js]: This looks inconsistent with the other files with how they declare `msg` from within the tests

File: tests/node/util/BsonTest.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const SDL = require('./../../../lib/js/dist/SDL.js');
+const Bson = SDL.util.Bson;
+const expect = require('chai').expect;
+
+
+describe('BsonTest', function () {
+    it('serialize deserialize test', function (done) {
```
Comment (nickschwab) [inline tests/node/util/BsonTest.js]: Test name does not match naming convention used in other tests. e.g. `testThingMethodBehavior`

File: tests/Validator.js [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+            }
+        } else {
+            return obj;
+        }
+        return result;
+    }
+
+    /**
+     * Validate an rpcMessage matches the given expectedParameters
+     * @param {RpcMessage} rpcMessage - ssageto validate against.
```
Comment (crokita) [inline tests/Validator.js]: Typo in the docs

File: tests/node/util/BitConverterTest.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+
+const expect = require('chai').expect;
+const SDL = require('./../../../lib/js/dist/SDL');
+const BitConverter = SDL.util.BitConverter;
+
+function assertEquals (val1, val2) {
+    expect(val1).to.be.deep.equal(val2);
+}
```
Comment (nickschwab) [inline tests/node/util/BitConverterTest.js]: Recurring feedback regarding assert definitions.

File: tests/node/rpc/messages/OnHmiStatusTests.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+const assertNull = Validator.assertNull.bind(Validator);
+const assertNullOrUndefined = Validator.assertNullOrUndefined.bind(Validator);
+const assertNotNull = Validator.assertNotNull.bind(Validator);
+const testNullBase = Validator.testNullBase.bind(Validator, 
+    FunctionID.keyForValue(FunctionID.OnHMIStatus), 
+    RpcType.NOTIFICATION);
+
+describe('OnHmiStatusTests', function () {
+    before(function () {
+        this.createMessage = function () {
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/OnHmiStatusTests.js]: Basically, this is the suggested way to extend tests for mocha. The BaseRpcTests.js is used to run the basic rpc tests and it is using the overridden createMessage.

File: tests/node/rpc/messages/RegisterAppInterfaceResponseTests.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+const Test = require('./../../../Test.js');
+const Validator = require('./../../../Validator.js');
+
+const assertTrue = Validator.assertTrue.bind(Validator);
+const assertEquals = Validator.assertEquals.bind(Validator);
+const assertNull = Validator.assertNull.bind(Validator);
+const assertNullOrUndefined = Validator.assertNullOrUndefined.bind(Validator);
+const assertNotNull = Validator.assertNotNull.bind(Validator);
+
+const testNullBase = Validator.testNullBase.bind(Validator,
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/RegisterAppInterfaceResponseTests.js]: fixed

File: tests/node/rpc/messages/RegisterAppInterfaceResponseTests.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        };
+    });
+
+    BaseRpcTests.tests();
+
+
+    it ('testRpcValues', function (done) {
+        let msg = this.msg;
+        // Test Values
+        const testSupportedDiagModes =  msg.getSupportedDiagModes();
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/RegisterAppInterfaceResponseTests.js]: I updated the order here to match the param order in the mobile rpc spec https://github.com/smartdevicelink/rpc_spec/blob/version/6_0_0/MOBILE_API.xml .  setSystemSoftwareVersion and setHMICapabilities were missing and so those were added to this now as well.

File: tests/node/rpc/messages/RegisterAppInterfaceTests.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+const Test = require('./../../../Test.js');
+const Validator = require('./../../../Validator.js');
+
+
+const assertTrue = Validator.assertTrue.bind(Validator);
+const assertEquals = Validator.assertEquals.bind(Validator);
+const assertNull = Validator.assertNull.bind(Validator);
+const assertNullOrUndefined = Validator.assertNullOrUndefined.bind(Validator);
+const assertNotNull = Validator.assertNotNull.bind(Validator);
+const testNullBase = Validator.testNullBase.bind(Validator,
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/RegisterAppInterfaceTests.js]: removed unused variables

File: tests/node/rpc/messages/RegisterAppInterfaceTests.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        assertTrue(Test.TRUE, Validator.validateTemplateColorScheme(Test.GENERAL_NIGHTCOLORSCHEME, testNightColorScheme));
+
+        // Invalid/Null Tests
+        rpcMessage = new RegisterAppInterface();
+        // TODO correlationId should be set automatically during construction using a correlation id generator.
+        rpcMessage.setCorrelationId(Test.GENERAL_INT);
+        assertNotNull(Test.NOT_NULL, rpcMessage);
+        testNullBase(rpcMessage);
+
+        assertNullOrUndefined(Test.NULL, rpcMessage.getSdlMsgVersion());
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/RegisterAppInterfaceTests.js]: I removed the the message for these assertion where it was a generic message. The error returned by default is accurate. For this one if there is an error it should say something like "line 116 expected null to not be null or undefined". The default messages were something I copied over from the java suite but they are not really relevant here. The custom error messages like 'Correlation ID should be defined.' are still there.

File: tests/node/rpc/structs/ImageTests.js (hunk 1 above)
Comment (russjohnson09) [reply tests/node/rpc/structs/ImageTests.js]: Added BaseStructTests which acts similarly to BaseRpcTests.

File: tests/Validator.js (hunk 3 above)
Comment (russjohnson09) [reply tests/Validator.js]: fixed typo

File: tests/node/rpc/enums/AppHMITypeTests.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
+
+const expect = require('chai').expect;
+const SDL = require('./../../../../lib/js/dist/SDL');
+
+const AppHMIType = SDL.rpc.enums.AppHMIType;
+
+function assertNotNull (msg, val) {
+    expect(val, msg).to.not.be.null;
+}
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/enums/AppHMITypeTests.js]: Using Validator static methods for all assertions.

File: tests/Validator.js [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+        const args = arguments;
+        let val,msg;
+        if (args.length === 2) {
+            val = args[1];
+            msg = args[0];
+        } else {
+            val = args[0];
+        }
+        expect(val, msg).not.to.be.null.and.not.to.be.undefined;
+    }
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/Validator.js]: fixed

File: tests/Validator.js [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+        if (args.length === 1) {
+            val1 = args[0];
+        } else if (args.length === 2) {
+            msg = args[0];
+            val1 = args[1];
+        } else {
+            throw new Error('Bad arguments');
+        }
+        expect(val1, msg).to.be.not.null;
+    }
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/Validator.js]: fixed

File: tests/node/rpc/messages/BaseRpcTests.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
+const SDL = require('./../../../../lib/js/dist/SDL.js');
+
+const RpcRequest = SDL.rpc.RpcRequest;
+const RpcResponse = SDL.rpc.RpcResponse;
+
+const Validator = require('./../../../Validator.js');
+
+const assertEquals = Validator.assertEquals.bind(Validator);
+const assertNotNullUndefined = Validator.assertNotNullUndefined.bind(Validator);
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/node/rpc/messages/BaseRpcTests.js]: fixed

File: tests/Validator.js [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+class Validator {
+    /**
+     * Takes an RpcStruct and converts it to a json object.
+     * @param {Object} obj Either a primative value, an array, or an object with getParameters defined. 
+     */
+    static getParametersJson (obj) {
+        let result;
+        if (typeof obj === 'object') {
+            if (typeof obj.getParameters === 'function') {
+                return this.getParametersJson(obj.getParameters());
```
[the remark this answers is not in the document]
Comment (russjohnson09) [reply tests/Validator.js]: fixed

File: tests/node/util/BsonTest.js (hunk 2 above)
Comment (russjohnson09) [reply tests/node/util/BsonTest.js]: testSerialize and testDeserialize used as test names now.

File: tests/node/util/BitConverterTest.js (hunk 4 above)
Comment (russjohnson09) [reply tests/node/util/BitConverterTest.js]: fixed

**Pull request merged**